lib/internal/test_runner/test.js

```diff
@@ -76,15 +77,38 @@ const { testNamePatterns, testOnlyFlag } = parseCommandLine();
 let kResistStopPropagation;
 
 function stopTest(timeout, signal) {
+  const deferred = createDeferredPromise();
+
   if (timeout === kDefaultTimeout) {
-    return once(signal, 'abort');
+    deferred[SymbolDispose] = addAbortListener(signal, deferred.resolve)[SymbolDispose];
+    return deferred;
+  }
+
+  const cleanAbort = addAbortListener(signal, throwOnAbort)[SymbolDispose];
+  const timer = setTimeout(() => {
+    deferred.resolve();
+    cleanAbort();
+  }, timeout);
+
+  timer.unref();
```

Comment:
Could/Should this be an `AbortSignal.any([signal, AbortSignal.timeout(timeout)])` etc?

Reply:
What do you mean about the timer not getting clean?
The `AbortSignal.any` uses weak refs, so once this is GCed it would clear the timeout... 

I think this suggestion could potentially reduce the complexity here, and also make it easier to land on v18, which doesn't have dispose & `addAbortListener` at the moment

WDYT?

Reply:
thank you! this is a much cleaner solution! the only problem here is that we create a new AbortSignal for every describe/test/hook run

Reply:
@MoLow what do you think? should we remove the addAbortListener and dispose in order to support v18 and v16?

Reply:
if we use `AbortSignal.any` and `AbortSignal.timeout`

the code will be:
```js
function stopTest(timeout, signal) {
  const deferred = createDeferredPromise();

  if (timeout === kDefaultTimeout) {
    deferred[SymbolDispose] = addAbortListener(signal, deferred.resolve)[SymbolDispose];
    return deferred;
  }

  const abortOrTimeoutSignal = AbortSignal.any([signal, AbortSignal.timeout(timeout)]);

  return {
    promise: PromisePrototypeThen(once(abortOrTimeoutSignal, 'abort'), () => {
      throw new ERR_TEST_FAILURE(
        `test timed out after ${timeout}ms`,
        kTestTimeoutFailure,
      );
    }),
    [SymbolDispose]: noop,
  };
}
```

and the reason we don't get max listeners warning is because `AbortSignal.any` create new abort signal every time and add listener to it

Reply:
Removed the use of `addAbortListener` and `SymbolDispose` so it will be easier to land on v18 and v16

Reply:
@rluvaton I think we can simplify it more by adding the timeout conditionally (but it would still mean creating additional AbortSignal which comes with a price)

Consider something similar to this:
`AbortSignal.any([signal, timeout !== kDefaultTimeout && AbortSignal.timeout(timeout)].filter(Boolean));`

Reply:
I don't think this actually simplifies 😬

Reply:
@rluvaton I has something like this in mind:

```javascript
function stopTest(timeout, signal) {
  const timeoutSignal = timeout !== kDefaultTimeout && AbortSignal.timeout(timeout);
  const composedSignal = AbortSignal.any([signal, timeoutSignal].filter(Boolean));
  return PromisePrototypeThen(once(composedSignal, 'abort'), (e) => {
    if (e.target.reason !== timeoutSignal.reason) {
      return;
    }
    throw new ERR_TEST_FAILURE(
      `test timed out after ${timeout}ms`,
      kTestTimeoutFailure,
    );
  })
};
```
But current solution works for me 🙂

Reply:
If we accept the extra creation of an abort signal I highly prefer that

Reply:
I assume we would want to profile both and decide based on that?

Reply:
I've created a benchmark PR and after it is merged we can check different solutions:
- https://github.com/nodejs/node/pull/48931